### ISSUE TITLE
[RFC] Generate ir.sequence.date_range when creating / updating account.fiscal.year

### ISSUE THREAD

**Issue by legalsylvain** [OCA/account-financial-tools#1003]
hi all, 

I'm working on Odoo CE 12.0, in a multicompany context. Each years, I have to create new fiscal years for each company. In the meantime, I have also to configure each ``ir.sequence`` for each journal, adding a new line ``ir.sequence.date_range``. It's quite fastidious and error prone.

I was thinking to automate the creation of that new lines, for each journals, when creating a new fiscal year.

In fact, I was also thinking to automate the creation of the fiscal years each years. ;-)
Do you have similar needs ? Are there some OCA modules that covers that features ? 

kind regards.

CC : original contributors of fiscal years module : @alexis-via, @eLBati, @MiquelRForgeFlow

![image](https://user-images.githubusercontent.com/3407482/83251993-ca5fa700-a1aa-11ea-9280-7b707d186b98.png)

**Comment by HaraldPanten** [comment]
I'm not sure if I understood you well, but I think that just checking "Use subsequences per date_range" you don't need to do anything else.

Try this:

With your current configuration (screenshot), create a new draft invoice and put invoice date in 2022. Then validate it.

If I'm not wrong, you invoice sequence will be: BILL/2022/0001. So... you don't need to create every year many subsequences for all your journals. You just need to add the subsequence, only if this subsequence is changing something compared to previous years, like starting from a different number from 1, for example.

Am I right?

**Comment by pedrobaeza** [comment]
FYI `account.fiscal.year` are not really needed, except if you don't have regular one year fiscal periods (which can be the starting one, but not the rest). If something is not working this way, it's because custom code is bad using these fiscal years. Example of "bad usage" (doing them mandatory): https://github.com/OCA/account-financial-tools/pull/870

Also, that lines are automatically populated when involved date belongs to other date range.

In summary, I don't see the need for this.

**Comment by legalsylvain** [comment]
Hi @pedrobaeza and @HaraldPanten, thanks for your answer.

>  except if you don't have regular one year fiscal periods (which can be the starting one, but not the rest).

There are 60 companies in my Odoo, and for many of them (like > 10), the fiscal year is not from january to december, but from october to september (for exemple), each year. So each year the setting should be ``date_from`` 20xx-10-01 ; ``date_end`` : 20xy-09-30. It is legal in France.

> Also, that lines are automatically populated when involved date belongs to other date range.

Indeed, I created a Bill in 2025-01-01 and here is the result :

(the result is wrong in my case, because for the companies with lagged fiscal year, the number should be ``BILL/24/0001``)

![image](https://user-images.githubusercontent.com/3407482/83255069-106b3980-a1b0-11ea-8e7a-eb206e82f530.png)

**Comment by HaraldPanten** [comment]
Hmmm... Then I think that "standard" doesn't work for you as you mentioned that fiscal years don't start on 01/01 and finish the 31/12.

What would be your purpose to solve this need, then?

I think it doesn't exist a specific feature to cover this case of use.

**Comment by pedrobaeza** [comment]
Standard works the same. The only problem for needing account.fiscal.year is a length different from the year. If not, you can select the last month of the fiscal year and the last day of that month, that are 2 fields in res.company.

**Comment by legalsylvain** [comment]
> Standard works the same. The only problem for needing account.fiscal.year is a length different from the year. If not, you can select the last month of the fiscal year and the last day of that month, that are 2 fields in res.company.

No it doesn' work. 
I set the following settings on ``res.company``

![image](https://user-images.githubusercontent.com/3407482/83258043-ddc43f80-a1b5-11ea-8f69-1a9e1fc101a8.png)

and when I create a bill (date = 2030-01-01), the following sub sequence created is still bad : 

![image](https://user-images.githubusercontent.com/3407482/83257979-bf5e4400-a1b5-11ea-8a35-569654d51da9.png)

And so, the number is bad.

![image](https://user-images.githubusercontent.com/3407482/83258073-ecaaf200-a1b5-11ea-8563-75e2a6c8a8f4.png)

Expected result : 
- sub sequence from 2029-10-01 to 2030-09-30
- Bill number : BILL/2029/0001

Or did I missed something ?

**Comment by legalsylvain** [comment]
> What would be your purpose to solve this need, then?

1. Propose a Fix / or make a OCA module that handles correctly the two values ``fiscalyear_last_day`` and ``fiscalyear_last_month`` when generating sub sequence.

2. Propose another module (no dependency with the first one) that autogenerate fiscal years but without generation of subsquences. (as there are automatically generated).

**Comment by pedrobaeza** [comment]
Yeah, indeed the sequences are for the whole year, but not that fiscal periods, but you need a whole new sequence infrastructure, because the prefix is got from the year, and in these cases, you need to manually indicate how to call them: "INV/2019-2020/001" for example?

**Comment by HaraldPanten** [comment]
> > Standard works the same. The only problem for needing account.fiscal.year is a length different from the year. If not, you can select the last month of the fiscal year and the last day of that month, that are 2 fields in res.company.
> 
> No it doesn' work.
> I set the following settings on `res.company`
> 
> ![image](https://user-images.githubusercontent.com/3407482/83258043-ddc43f80-a1b5-11ea-8f69-1a9e1fc101a8.png)
> 
> and when I create a bill (date = 2030-01-01), the following sub sequence created is still bad :
> 
> ![image](https://user-images.githubusercontent.com/3407482/83257979-bf5e4400-a1b5-11ea-8a35-569654d51da9.png)
> 
> And so, the number is bad.
> 
> ![image](https://user-images.githubusercontent.com/3407482/83258073-ecaaf200-a1b5-11ea-8563-75e2a6c8a8f4.png)
> 
> Expected result :
> 
> * sub sequence from 2029-10-01 to 2030-09-30
> * Bill number : BILL/2029/0001
> 
> Or did I missed something ?

I think that subsequence is not prepared for this case of use. I agree that in this case of use something has to be developed

**Comment by HaraldPanten** [comment]
> > What would be your purpose to solve this need, then?
> 
> 1. Propose a Fix / or make a OCA module that handles correctly the two values `fiscalyear_last_day` and `fiscalyear_last_month` when generating sub sequence.
> 2. Propose another module (no dependency with the first one) that autogenerate fiscal years but without generation of subsquences. (as there are automatically generated).

I'd prefer a new module, as I think that this functionality is not that common (depending on accounting laws and/or industries).

**Comment by legalsylvain** [comment]
>  you need to manually indicate how to call them: "INV/2019-2020/001" for example?

Well, the text included is another topic : 
My accountant supposed that a fiscal year that begin from 2019-10-01 to 2020-09-30 should have numbers like ``Bill/19/000x`` (because otherwise, it's a mess to write)


but yes, another idea could be to have another custom key in the prefix / suffix values, like : 

![image](https://user-images.githubusercontent.com/3407482/83259433-53c9a600-a1b8-11ea-9300-e9eac01daeeb.png)

If yes, for a range from 2019-10-01 to 2020-09-30, It could generate a BILL/19-20/xxx (or whatever)

**Comment by pedrobaeza** [comment]
You can use a specific placeholder indeed (or give several possibilities), then override the proper method and will need to find the starting and ending dates for the fiscal year (no matter if specified from `account.fiscal.year` or by last month/day), and put variables in the context for being able to replace the placeholder.

**Comment by legalsylvain** [comment]
Closing.
Solved by https://github.com/OCA/account-financial-tools/pull/1005
And https://github.com/OCA/account-financial-tools/pull/1148

Thanks for your contribution.
